pyiron_base/generic/object.py

```diff
@@ -0,0 +1,93 @@
+# coding: utf-8
+# Copyright (c) Max-Planck-Institut für Eisenforschung GmbH - Computational Materials Design (CM) Department
+# Distributed under the terms of "New BSD License", see the LICENSE file.
+
+"""
+The core class in pyiron, linking python to the database to file storage.
+"""
+
+from abc import ABC
+from pyiron_base import DataContainer
+from pyiron_base.settings.generic import Settings
+
+__author__ = "Liam Huber"
+__copyright__ = (
+    "Copyright 2020, Max-Planck-Institut für Eisenforschung GmbH - "
+    "Computational Materials Design (CM) Department"
+)
+__version__ = "0.0"
+__maintainer__ = "Liam Huber"
+__email__ = "huber@mpie.de"
+__status__ = "development"
+__date__ = "Mar 23, 2021"
+
+s = Settings()
+
+
+class HasStorage(ABC):
```

Comment:
To help "super-users" creating new classes, I want a class that saves its attributes automatically.  This came up with @Leimeroth in pyiron_contrib for example.  This would work nicely with `HasStorage` since we could just route `__getattr__`/`__setattr__` to `self.storage`.  I could add this here or in a separate class, thoughts?

Reply:
@pmrv , my trite answer is that we already have this class -- it's `DataContainer` and it's beautiful.

More seriously, I guess you mean such a class that *also* has the database functionality, e.g. to be logged there and referenced by other pyiron objects later.

I definitely like the direction you're going with the setters. I don't want this to be the standard behaviour, because there are many times we *don't* want an attribute to be serialized, e.g. consider `job.project` where we really don't want to be serializing the project repeatedly with every job. And we see already in `DataContainer` that once serialization is the default getting around it is a pain. However, I see it could be useful in a bunch of cases.

I think we'll probably need to add more bells and whistles than just the (ge)setters at the end of the day, but it's a good start. What about something like this:

```python
class PyironData(PyironObject):
  def __getattr__(self, key):
    return self.storage[key]

  def __setattr__(self, key, value):
    return self.storage[key] = value
```

It's just not totally clear to me how this sort of magic method overwrite will play out in the init of the super classes where we do stuff like `self._storage = DataContainer(table_name='storage')`.

Another option is to brute force it by putting wrappers around *each* field we care about:

```python
class MyData(PyironObject):
  def __init__(self, *args, **kwargs):
    super().__init__()
    self.storage._field1 = 'foo'
    self.storage._field2 = 42

  @property 
  def field1(self):
    return self.storage._field1

  @field1.setter 
  def field1(self, val):
    self.storage._field1 = val

  @property 
  def field2(self):
    return self.storage._field2

  @field2.setter 
  def field2(self, val):
    self.storage._field2 = val
```

If we have to go the second route, it would be nice to have some sort of wrapper to procedurally add these getters and setters so we can write something very roughly like

```python
@expose_storage('field1', 'field2')
class MyData(PyironObject):
  def __init__(self, *args, **kwargs):
    super().__init__()
    self.storage._field1 = 'foo'
    self.storage._field2 = 42
```

Anyhow, as Jan already approved this version of the PR I'll merge it later tonight and we can continue these data ideas in a new PR 😉